src/suas_planning/src/GlobalWaypointPlanner.cpp

```diff
@@ -1,28 +1,141 @@
+            std::size_t h2 = (node.y_ + 0x7261735f) >> 1;
+            std::size_t h3 = (node.action_ + 0x7261735f) >> 1;
+            std::size_t temp = (h1 ^ h2) + (h1 << 6) + (h1 >> 2);
+            return (temp ^ h3) + (temp << 6) + (h3 >> 2);
+        }
+    };
+
+    template<> struct equal_to<suas_planning::GlobalWaypointPlanner::Node> {
+        bool operator()(
+            const suas_planning::GlobalWaypointPlanner::Node& lhs,
+            const suas_planning::GlobalWaypointPlanner::Node& rhs
+        ) const noexcept {
+            bool x = lhs.x_ == rhs.y_;
+            bool y = lhs.y_ == rhs.y_;
+            bool action = lhs.action_ == rhs.action_;
+            bool from = lhs.from_ == rhs.from_;
+            bool goal = lhs.goal_ == rhs.goal_;
+            return x && y && action && from && goal;
+        }
+    };
+    
+} // namespace std for std::hash injection
+
 namespace suas_planning {
 
+static const int NO_ACTION = -1;
+static const int NUM_ACTIONS = 8;
+static const int dX[NUM_ACTIONS] = {0, 0, 1, 1, 1, -1, -1, -1};
+static const int dY[NUM_ACTIONS] = {1, -1, 0, 1, -1, 0, 1, -1};
+static const double action_cost[NUM_ACTIONS] = {1, 1, 1, M_SQRT2, M_SQRT2, 1, M_SQRT2, M_SQRT2};
+
 GlobalWaypointPlanner::GlobalWaypointPlanner(
-        ros::NodeHandle& nh,
-        double start_x,
-        double start_y,
-        double goal_x,
-        double goal_y,
-        std::vector<Obstacle> obstacles,
-        std::vector<Waypoint> waypoints,
-        double radius=1.0,
-        double resolution=1.0) :
-            nh_(nh),
-            vehicle_radius_(radius),
-            resolution_(resolution),
-            current_map_(ComputeMapSize(obstacles)),
-            waypoints_(waypoints),
-            map_meta_(&current_map_, ComputeMapWidth(obstacles), ComputeMapHeight(obstacles)),
-            start_(start_x, start_y),
-            goal_(goal_x, goal_y) {
-        UpdateMap(obstacles);
+    ros::NodeHandle& nh,
+    double start_x,
+    double start_y,
+    double goal_x,
+    double goal_y,
+    std::vector<Obstacle> obstacles,
+    std::vector<Waypoint> waypoints,
+    double radius,
+    double resolution) :
+        nh_(nh),
+        current_map_(ComputeMapSize(obstacles)),
+        waypoints_(waypoints),
+        map_meta_(current_map_, ComputeMapWidth(obstacles), ComputeMapHeight(obstacles)),
+        start_((int) round(start_x), (int) round(start_y)),
+        goal_((int) round(goal_x), (int) round(goal_y)),
+        vehicle_radius_(radius),
+        resolution_(resolution) {
+    UpdateMap(obstacles);
+}
+
+/*
+ * Generates a route/plan for waypoints and goals
+ */
+std::vector<std::string> GlobalWaypointPlanner::GeneratePlan() {
+    Waypoint current_start = start_;
+    std::queue<Waypoint, std::deque<Waypoint>> sub_goals;
+    std::vector<std::string> sub_plans;
+    if (!sub_goals.size()) {
+        for (std::vector<Waypoint>::iterator waypoint_it = waypoints_.begin(); waypoint_it != waypoints_.end(); waypoint_it++) {
```

Comment:
```cpp
for(Waypoint& waypoint : waypoints) {}
```

 does this same exact thing in a more concise syntax. Access your elements using the `waypoint` variable as if it were a normal variable. The & guarantees a reference to the object so  no copy, and modifying `waypoint` in this scope will modify the element in `waypoints`. If no modifications are necessary, throw a `const` in front of the declaration such as `const Waypoint& waypoint`.

Reply:
Resolved in 64e1edb